File: bankAccount/20121206-adamsky/src/main/java/com/novoda/dojos/bankaccount/Account.java
package com.novoda.dojos.bankaccount;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Account {

    private final List<Transaction> transactions;

    public Account() {
        transactions = Collections.emptyList();
    }

    private Account(Transaction head, List<Transaction> tail) {
        transactions = new ArrayList<Transaction>(tail);
        transactions.add(head);
    }

    public Account deposit(Money deposit) {
        return action(new Deposit(deposit));
    }

    public Money balance() {
        Money tmp = new Money();
        for (Transaction t : transactions) {
            tmp = getNext(tmp, t);
        }
        return tmp;
    }

    private Money getNext(Money tmp, Transaction t) {
        if (t instanceof Deposit) {
            tmp = tmp.add(t.amount());
        } else if (t instanceof Withdraw) {
            tmp = tmp.sub(t.amount());
        }
        return tmp;
    }

    public Account withdraw(Money withdrawal) {
        return action(new Withdraw(withdrawal));
    }

    private Account action(Transaction transaction) {
        return new Account(transaction, transactions);
    }
}

File: bankAccount/20181019-juanky-xavi/src/main/java/com/novoda/dojos/bankaccount/Account.kt
package com.novoda.dojos.bankaccount

class Account {

    private val ledger: Ledger = Ledger()

    fun balance(): Balance = ledger.balance()

    fun deposit(amount: Amount) {
        ledger.deposit(amount)
    }

    fun withdraw(amount: Amount) {
        ledger.withdraw(amount)
    }

    fun transfer(amount: Amount, toAccount: Account) {
        ledger.withdraw(amount)
        toAccount.deposit(amount)
    }
}

data class Balance(val amount: Amount)

data class Ledger(val transactions: MutableList<Transaction> = mutableListOf()) {
    fun deposit(amount: Amount) {
        transactions.add(Deposit(amount))
    }

    fun withdraw(amount: Amount) {
        transactions.add(Withdrawal(amount))
    }

    fun balance(): Balance = Balance(transactions.fold(Amount(0)) { acc, transaction ->
        transaction.applyTo(acc)
    })
}

data class Amount(val value: Int) {
    operator fun plus(other: Amount) = Amount(value + other.value)
    operator fun minus(other: Amount) = Amount(value - other.value)
}

sealed class Transaction {
    abstract fun applyTo(amount: Amount): Amount
}

data class Deposit(val transactionAmount: Amount) : Transaction() {
    override fun applyTo(amount: Amount): Amount = amount + transactionAmount
}

data class Withdrawal(val transactionAmount: Amount) : Transaction() {
    override fun applyTo(amount: Amount): Amount = amount - transactionAmount
}
